stored_requests/backends/db_provider/mysql_dbprovider.go

```diff
@@ -75,7 +86,97 @@ func (provider *MySqlDbProvider) ConnString() string {
 		buffer.WriteString(provider.cfg.Database)
 	}
 
-	return buffer.String()
+	queryStr, err := provider.generateQueryString()
+	if err != nil {
+		return "", err
+	}
+	if queryStr != "" {
+		buffer.WriteString("?")
+		buffer.WriteString(queryStr)
+	}
+
+	return buffer.String(), nil
+}
+
+func setupTLSConfig(provider *MySqlDbProvider) error {
+	rootCertPool := x509.NewCertPool()
+
+	pem, err := ioutil.ReadFile(provider.cfg.TLS.RootCert)
+	if err != nil {
+		return err
+	}
+
+	if ok := rootCertPool.AppendCertsFromPEM(pem); !ok {
+		return err
+	}
+
+	var clientCert []tls.Certificate
+	if provider.cfg.TLS.ClientCert != "" && provider.cfg.TLS.ClientKey != "" {
+		clientCert = make([]tls.Certificate, 0, 1)
+		certs, err := tls.LoadX509KeyPair(provider.cfg.TLS.ClientCert, provider.cfg.TLS.ClientKey)
+		if err != nil {
+			return err
+		}
+
+		clientCert = append(clientCert, certs)
+	}
+
+	mysql.RegisterTLSConfig(customTLSKey, &tls.Config{
+		RootCAs:               rootCertPool,
+		Certificates:          clientCert,
+		InsecureSkipVerify:    true,
+		VerifyPeerCertificate: verifyPeerCertFunc(rootCertPool),
+	})
+
+	return nil
+}
+
+// verifyPeerCertFunc returns a function that verifies the peer certificate is
+// in the cert pool.
+func verifyPeerCertFunc(pool *x509.CertPool) func([][]byte, [][]*x509.Certificate) error {
+	return func(rawCerts [][]byte, _ [][]*x509.Certificate) error {
+		if len(rawCerts) == 0 {
+			return errors.New("no certificates available to verify")
+		}
+
+		cert, err := x509.ParseCertificate(rawCerts[0])
+		if err != nil {
+			return err
+		}
+
+		opts := x509.VerifyOptions{Roots: pool}
+		if _, err = cert.Verify(opts); err != nil {
+			return err
+		}
+		return nil
+	}
+}
+
+func (provider *MySqlDbProvider) generateQueryString() (string, error) {
+	tls := ""
+
+	if provider.cfg.TLS.RootCert != "" {
+		tls = "true"
+
+		if err := setupTLSConfig(provider); err != nil {
+			return "", err
+		}
+
+		if provider.cfg.TLS.ClientCert != "" && provider.cfg.TLS.ClientKey != "" {
+			tls = customTLSKey
+		}
+	}
+
+	if tls != "" {
+		if len(provider.cfg.QueryString) == 0 {
+			return "tls=" + tls, nil
+		}
+		if !strings.Contains(provider.cfg.QueryString, "tls=") {
+			return "tls=" + tls + "&" + provider.cfg.QueryString, nil
+		}
+	}
+
+	return provider.cfg.QueryString, nil
```

Comment:
Please refer to thiscomment: https://github.com/prebid/prebid-server/pull/2507#discussion_r1114716636 
`tls` can only be `customTLSKey`.

Reply:
@VeronikaSolovei9 thank you for pointing out this missed case.
Fixed and added corresponding tests.

Reply:
@victorshevtsov I think there still might be a hole in the logic. I don't think that it is necessarily true that `tls` can only ever be `customTLSKey`. I think that if `tls` is defined in the `query_string` we should use that value instead of using the default `customTLSKey`. The version of `generateQueryString` function I originally shared was written with that intent. The issue I see is that the first parameter passed to `RegisterTLSConfig` must match the constructed query string `tls` value and that may either be the `tls` value provided in `query_string` by the host or `customTLSKey` if `tls` was not provided in `query_string` by the host.

In that case, I think some adjustment to `generateQueryString` might fix this (I didn't test this):
```
func (provider *MySqlDbProvider) generateQueryString() (string, error) {
        // parse the query string
	queryParams, err := url.ParseQuery(provider.cfg.QueryString)
	if err != nil {
		return "", err
	}
        // determine if TLS is in the query string and get the value
	tlsValues, tlsInQueryStr := queryParams["tls"]
	tls := ""
	if len(tlsValues) > 1 {
		return "", error(TODO)
	}
	if len(tlsValues) == 1 {
		tls = tlsValues[0]
	}

        // if there is a root cert we know we want to use TLS
	if provider.cfg.TLS.RootCert != "" {
                // no TLS was found in the query string so use the default key
		if !tlsInQueryStr {
			tls = customTLSKey
		}

                 // pass the key (either from query string or the default) down to set up TLS
		if err := setupTLSConfig(provider, tls); err != nil {
			return "", err
		}
	}

	if tls != "" {
		if len(provider.cfg.QueryString) == 0 {
			return "tls=" + tls, nil
		}
		if !tlsInQueryStr {
			return "tls=" + tls + "&" + provider.cfg.QueryString, nil
		}
	}

        // I think this line would have returned the wrong string in the current code because if tls != "" but there was a query string already containing the key tls, we would have registered with customTLSKey but returned a string containing whatever tls was put in the query string by the host
	return provider.cfg.QueryString, nil
}
```

Thoughts?

Reply:
How should we handle a case If the tls option is passed by the query string it has a special meaning as describe in the docs?
![image](https://user-images.githubusercontent.com/32890592/221140053-f49562ad-b737-4f54-a509-40a4e931a6d2.png)

Reply:
Does PBS have to do anything special in these cases? It didn't seem like it to me. For the invalid / self-signed cert case it's on the host to correctly generate that and provide it as a root cert to PBS. The initial ping by PBS should fail if they set it up incorrectly.
Wouldn't these values also be passed as the first param to `RegisterTLSConfig`?
Are you just referring to validation of those values? I don't think I would worry about that because technically any name is valid.

Reply:
Done.